File: dotnet/src/Downstairs.Infrastructure/Persistence/Models/AuthenticationLog.cs
namespace Downstairs.Infrastructure.Persistence.Models;

public partial class AuthenticationLog
{
    public long Id { get; set; }

    public string AuthenticatableType { get; set; } = null!;

    public long AuthenticatableId { get; set; }

    public string? IpAddress { get; set; }

    public string? UserAgent { get; set; }

    public DateTime? LoginAt { get; set; }

    public bool LoginSuccessful { get; set; }

    public DateTime? LogoutAt { get; set; }

    public bool ClearedByUser { get; set; }

    public string? Location { get; set; }
}

File: php-laraval-app/app/Models/AuthenticationLog.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Relations\MorphTo;

class AuthenticationLog extends Model
{
    public $timestamps = false;

    protected $table = 'authentication_log';

    protected $fillable = [
        'ip_address',
        'user_agent',
        'login_at',
        'login_successful',
        'logout_at',
        'cleared_by_user',
        'location',
    ];

    protected $casts = [
        'cleared_by_user' => 'boolean',
        'location' => 'array',
        'login_successful' => 'boolean',
        'login_at' => 'datetime',
        'logout_at' => 'datetime',
    ];

    /**
     * Define the alias of the columns or relations.
     *
     * @var array<string,string>
     */
    protected array $aliases = [
        'user' => 'authenticatable',
        'user_id' => 'authenticatable_id',
    ];

    /**
     * Define the models for polymorphic relations.
     *
     * @var array<string,string[]>
     */
    protected array $relationsMorphMap = [
        'authenticatable' => [User::class],
    ];

    public function __construct(array $attributes = [])
    {
        if (! isset($this->connection)) {
            $this->setConnection(config('authentication-log.db_connection'));
        }

        parent::__construct($attributes);
    }

    public function getTable()
    {
        return config('authentication-log.table_name', parent::getTable());
    }

    public function authenticatable(): MorphTo
    {
        return $this->morphTo()->withTrashed();
    }
}
